docs/prefetch.md

```diff
@@ -0,0 +1,112 @@
+- The prefetch context, that holds the state with a bunch o objects that are important for the prefetch logic
+- On RenderProvider when a navigation is triggered, we go to the prefetch cache to see if there is data there.
+
+### The Prefetch context
+
+This is the prefetch state interface:
+```
+
+interface PrefetchState {
+  routesCache: LRUCache<PrefetchRouteData>
+  pathsCache: {
+    other: LRUCache<PrefetchCacheObject>
+    product: LRUCache<PrefetchCacheObject>
+    search: LRUCache<PrefetchCacheObject>
+  }
+  pathsState: Record<string, PathState>
+  routePromise: Record<string, RoutePromise | null>
+  queue: PQueue
+}
+```
+
+- routesCache: A LRUCache that holds data for each page (can be pageCache if we want to change). The key is the page id i.e.: 'store.product', 'store.search#category'. Each entry has: the extensions for that page, the components for that page and the messages for that page.
+- pathsCache: an object with three LRU caches. One for products, one for search routes and one for any other of page, like landing pages. Each entry has as key the path to it like: '/banana/p', '/department'. The value it stores is an object with: routeId (the page id, like store.product), matchingPage (a data that contains a bunch of information that vary by path and are created by render-server and pages), contentResponse (this holds content, component and messages datas for pages that have unique content for them, will be null if not applicable) and queryData that contains the query data (like product query result, search query result) for that route.
+- pathsState: a map that stores data for each path. The key is the path (i.e.: /banana/p). It helps storing quickly the routeId for that path and the status of the request, to prevent concurrency issues.
+- routePromise: a map that stores the state of the request of that route/page. The key is the page id (i.e.: store.product). Prevents concurrency issues, like two requests to the same page being done at the same time.
+- queue: a promise queue, that allows up to 5. Simultaneous requests promises (each `<Link>` generates only one entry).
+
+In all of these LRU caches, the max age of the data saved is 30min. This prevents users from navigating, leaving the tab opened, then resuming the navigation and seeing really old data.
+They have different sizes, but no greater than 100. 
+
+#### PrefetchContextProvider
+
+It wraps all components, on start of the runtime it sets up the queue but leaves if turned off. After receiving the load signal from the browser, it waits a few extra seconds, then turns on the queue and the requests starts. We had to do it because otherwise, the donwload of prefetched components was affecting lighthouse scores.
+It also waits for navigation events, when they happen, we clear the queue, and pause it for one second to allow the browser to have some time to request assets and have more free network and connections.
```

Comment:
```suggestion
It wraps all components, on the start of the runtime, it sets up the queue but leaves if turned off. After receiving the load signal from the browser, it waits a few extra seconds, then turns on the queue and the requests start. We had to do it because otherwise, the download of prefetched components was affecting lighthouse scores.
It also waits for navigation events, when they happen, we clear the queue, and pause it for one second to allow the browser to have some time to request assets and have more free network and connections.
```